### PR TITLE
btl/tcp: ignore IPV6 link local addresses

### PR DESCRIPTION
**Pull request by ggouaillardet**
Refs open-mpi/ompi#585

### CODE REVIEW

**Comment by mellanox-github** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/jenkins/job/gh-ompi-master-pr/529/

**Build Log**
*last 50 lines*

 ```
[...truncated 6307 lines...]
libtool: install: /usr/bin/install -c .libs/mca_btl_usnic.lai /var/lib/jenkins/jobs/gh-ompi-master-pr/workspace/ompi_install1/lib/openmpi/mca_btl_usnic.la
libtool: finish: PATH="/hpc/local/bin::/usr/local/bin:/bin:/usr/bin:/usr/sbin:/hpc/local/bin:/hpc/local/bin/:/hpc/local/bin/:/sbin:/usr/sbin:/bin:/usr/bin:/usr/local/sbin:/opt/ibutils/bin:/sbin" ldconfig -n /var/lib/jenkins/jobs/gh-ompi-master-pr/workspace/ompi_install1/lib/openmpi
----------------------------------------------------------------------
Libraries have been installed in:
   /var/lib/jenkins/jobs/gh-ompi-master-pr/workspace/ompi_install1/lib/openmpi

If you ever happen to want to link against installed libraries
in a given directory, LIBDIR, you must either use libtool, and
specify the full pathname of the library, or use the `-LLIBDIR'
flag during linking and do at least one of the following:
   - add LIBDIR to the `LD_LIBRARY_PATH' environment variable
     during execution
   - add LIBDIR to the `LD_RUN_PATH' environment variable
     during linking
   - use the `-Wl,-rpath -Wl,LIBDIR' linker flag
   - have your system administrator add LIBDIR to `/etc/ld.so.conf'

See any operating system documentation about shared libraries for
more information, such as the ld(1) and ld.so(8) manual pages.
----------------------------------------------------------------------
make[3]: Leaving directory `/scrap/jenkins/jenkins/jobs/gh-ompi-master-pr/workspace/opal/mca/btl/usnic'
make[2]: Leaving directory `/scrap/jenkins/jenkins/jobs/gh-ompi-master-pr/workspace/opal/mca/btl/usnic'
Making install in mca/btl/tcp
make[2]: Entering directory `/scrap/jenkins/jenkins/jobs/gh-ompi-master-pr/workspace/opal/mca/btl/tcp'
  CC       btl_tcp.lo
  CC       btl_tcp_component.lo
  CC       btl_tcp_endpoint.lo
  CC       btl_tcp_frag.lo
  CC       btl_tcp_proc.lo
  CC       btl_tcp_ft.lo
btl_tcp_proc.c: In function 'mca_btl_tcp_retrieve_local_interfaces':
btl_tcp_proc.c:358: error: implicit declaration of function 'opal_net_islinklocal'
make[2]: *** [btl_tcp_proc.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory `/scrap/jenkins/jenkins/jobs/gh-ompi-master-pr/workspace/opal/mca/btl/tcp'
make[1]: *** [install-recursive] Error 1
make[1]: Leaving directory `/scrap/jenkins/jenkins/jobs/gh-ompi-master-pr/workspace/opal'
make: *** [install-recursive] Error 1
Build step 'Execute shell' marked build as failure
TAP Reports Processing: START
Looking for TAP results report in workspace using pattern: **/*.tap
Did not find any matching files.
Anchor chain: could not read file with links: /var/lib/jenkins/jobs/gh-ompi-master-pr/workspace/jenkins_sidelinks.txt (No such file or directory)
[copy-to-slave] The build is taking place on the master node, no copy back to the master will take place.
Setting commit status on GitHub for https://github.com/open-mpi/ompi/commit/9f86134c971119cd165aae5c72b48d4d506ffa2e
[BFA] Scanning build for known causes...
[BFA] No failure causes found
[BFA] Done. 0s
Setting status of 025beb166ca93ecbaffbd4f4a4dc13d2541ed155 to FAILURE with url http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr/529/ and message: Build finished.
```

Test FAILed.

**Comment by mellanox-github** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/jenkins/job/gh-ompi-master-pr/530/

**Build Log**
*last 50 lines*

 ```
[...truncated 6307 lines...]
libtool: install: /usr/bin/install -c .libs/mca_btl_usnic.lai /var/lib/jenkins/jobs/gh-ompi-master-pr/workspace/ompi_install1/lib/openmpi/mca_btl_usnic.la
libtool: finish: PATH="/hpc/local/bin::/usr/local/bin:/bin:/usr/bin:/usr/sbin:/hpc/local/bin:/hpc/local/bin/:/hpc/local/bin/:/sbin:/usr/sbin:/bin:/usr/bin:/usr/local/sbin:/opt/ibutils/bin:/sbin" ldconfig -n /var/lib/jenkins/jobs/gh-ompi-master-pr/workspace/ompi_install1/lib/openmpi
----------------------------------------------------------------------
Libraries have been installed in:
   /var/lib/jenkins/jobs/gh-ompi-master-pr/workspace/ompi_install1/lib/openmpi

If you ever happen to want to link against installed libraries
in a given directory, LIBDIR, you must either use libtool, and
specify the full pathname of the library, or use the `-LLIBDIR'
flag during linking and do at least one of the following:
   - add LIBDIR to the `LD_LIBRARY_PATH' environment variable
     during execution
   - add LIBDIR to the `LD_RUN_PATH' environment variable
     during linking
   - use the `-Wl,-rpath -Wl,LIBDIR' linker flag
   - have your system administrator add LIBDIR to `/etc/ld.so.conf'

See any operating system documentation about shared libraries for
more information, such as the ld(1) and ld.so(8) manual pages.
----------------------------------------------------------------------
make[3]: Leaving directory `/scrap/jenkins/jenkins/jobs/gh-ompi-master-pr/workspace/opal/mca/btl/usnic'
make[2]: Leaving directory `/scrap/jenkins/jenkins/jobs/gh-ompi-master-pr/workspace/opal/mca/btl/usnic'
Making install in mca/btl/tcp
make[2]: Entering directory `/scrap/jenkins/jenkins/jobs/gh-ompi-master-pr/workspace/opal/mca/btl/tcp'
  CC       btl_tcp.lo
  CC       btl_tcp_component.lo
  CC       btl_tcp_endpoint.lo
  CC       btl_tcp_frag.lo
  CC       btl_tcp_proc.lo
  CC       btl_tcp_ft.lo
btl_tcp_proc.c: In function 'mca_btl_tcp_retrieve_local_interfaces':
btl_tcp_proc.c:358: error: implicit declaration of function 'opal_net_islinklocal'
make[2]: *** [btl_tcp_proc.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory `/scrap/jenkins/jenkins/jobs/gh-ompi-master-pr/workspace/opal/mca/btl/tcp'
make[1]: *** [install-recursive] Error 1
make[1]: Leaving directory `/scrap/jenkins/jenkins/jobs/gh-ompi-master-pr/workspace/opal'
make: *** [install-recursive] Error 1
Build step 'Execute shell' marked build as failure
TAP Reports Processing: START
Looking for TAP results report in workspace using pattern: **/*.tap
Did not find any matching files.
Anchor chain: could not read file with links: /var/lib/jenkins/jobs/gh-ompi-master-pr/workspace/jenkins_sidelinks.txt (No such file or directory)
[copy-to-slave] The build is taking place on the master node, no copy back to the master will take place.
Setting commit status on GitHub for https://github.com/open-mpi/ompi/commit/1cea09707046ab06cb013b0816a4344a9c965c9d
[BFA] Scanning build for known causes...
[BFA] No failure causes found
[BFA] Done. 0s
Setting status of 7e1aecb2e213f4533e404c457779405e60ce88a7 to FAILURE with url http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr/530/ and message: Build finished.
```

Test FAILed.

**Comment by mellanox-github** [comment]
* Coverity found 911 errors for all: http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr//ws/cov_build/all_531/output/errors/index.html

* Coverity found 5 errors for oshmem: http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr//ws/cov_build/oshmem_531/output/errors/index.html

* Coverity found 2 errors for fca: http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr//ws/cov_build/fca_531/output/errors/index.html <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by mellanox-github** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/jenkins/job/gh-ompi-master-pr/531/
Test PASSed.

File: opal/mca/btl/tcp/btl_tcp_proc.c [hunk 1]
```diff
@@ -355,6 +355,10 @@ static mca_btl_tcp_interface_t** mca_btl_tcp_retrieve_local_interfaces(void)
             if (6 == mca_btl_tcp_component.tcp_disable_family) {
                 continue;
             }
+            if (opal_net_islinklocal(&local_addr)) {
```
Comment (jsquyres) [inline opal/mca/btl/tcp/btl_tcp_proc.c]: Per the comment below, if opal_net_islinklocal() grows to understand IPv4, we might as well move this check outside of the switch/case block and do it for both IPv4 and IPv6.

**Comment by bosilca** [comment]
With the proposed solution we handle IPV4 and IPV6 differently. Today we exclude the IPV4 local interface using the btl_tcp_if_exclude. A more consistent solution might be a better approach, thus we should handle IPV6 in the same way to handle IPV4.

**Comment by jsquyres** [comment]
Mmm.  Good point.

However, we've run into confusion with ignoring the loopback interface via if_include/exclude before (e.g., people exclude "eth3", but forget to *also* exclude "lo0").  Is it time to have a new boolean MCA param (btl_tcp|oob_tcp)_ignore_local_only_interfaces that will cause the TCP OOB/BTL to ignore loopback and link-local interfaces?

**Comment by ggouaillardet** [comment]
mmm too :-)

i did make a special case for IPV6 link local addresses on purpose.

local interface and link-local addresses are two different things :
- it is impossible to use the local interface to communicate with a remote host
- it is possible (with some restrictions) to use link-local addresses to communicate with a remote host


From [Wikipedia](http://http://en.wikipedia.org/wiki/Link-local_address)

> In [RFC 3927](http://tools.ietf.org/html/rfc3927), the Internet Engineering Task Force has reserved > the address block 169.254.1.0 through 169.254.254.255 for link-local addressing in Internet Protocol Version 4. Link-local addresses are assigned to interfaces by host-internal, i.e. stateless, address autoconfiguration when other means of address assignment are not available.

> In the Internet Protocol Version 6 (IPv6), the address block fe80::/10 has been reserved for link-local unicast addressing. The actual link local addresses are assigned with the prefix fe80::/64. They may be assigned by automatic (stateless) or stateful (e.g. manual) mechanisms.

and here is the relevant parts :

> [RFC 3927](http://tools.ietf.org/html/rfc3927) warns against the simultaneous use of IPv4 addresses of different scope, such as configuring link-local addresses as well as globally routable addresses on the same interface. Therefore, hosts search for a DHCP server on the network before assigning link-local addresses.

> Unlike IPv4, IPv6 requires a link-local address to be assigned to every network interface on which the IPv6 protocol is enabled, even when one or more routable addresses are also assigned.
>Consequently, IPv6 hosts usually have more than one IPv6 address assigned to each of their IPv6-enabled network interfaces. The link-local address is required for IPv6 sublayer operations of the Neighbor Discovery Protocol, as well as for some other IPv6-based protocols, like DHCPv6.

my point is :
- for IPv4 : link-local addresses are likely a config error, so we can simply blame the enduser/sysadmins for using them
- for IPv6 : link-local addresses are mandatory, and we must handle them appropriately.

why don't we first decide how to handle link-local addresses once for all first :
- IPv4 vs IPv6 ?
- should we simply discard them, period ?
- should we handle them in OPAL, and let each component (btl/tcp, oob/tcp) decide what to do ?

FWIW, i could not find a IN6_IS_ADDR_LINKLOCAL like macro for IPv4.

As far as i am concerned, i am fine with discarding link-local addresses for both IPv4 and IPv6 at the
OPAL layer. That being said, i do not think i should make this decision alone.

any thoughts ? shall i issue a RFC on the devel ML ?
out of curiosity, are you aware of sites using OpenMPI with IPv6 ?

**Comment by bosilca** [comment]
I always fall back to TCP when I have any doubts about a BTLs (and many other internal components), and as a lot of times I'm working on my laptop I use the local interface. So I would like to keep the definition of what local means local to the BTL, and don't prevent local TCP connections too deep in the software stack.

**Comment by hppritcha** [comment]
Gilles, I suggest an RFC on the devel ML route.

**Comment by ggouaillardet** [comment]
per the RFC, i will make a new PR and close this one

**Pull request closed without merging**